packages/notifier/src/types.js

```diff
@@ -36,7 +43,62 @@
 // /////////////////////////////////////////////////////////////////////////////
 
 /**
- * @typedef {number | undefined} UpdateCount a value used to mark the position
+ * @template T
+ * @typedef {object} PublicationRecord
+ * @property {IteratorResult<T>} head
+ * @property {bigint} publishCount
+ * @property {PublicationList<T>} tail
+ */
+
+/**
+ * @template T
+ * @typedef {ERef<PublicationRecord<T>>} PublicationList
+ * Will be shared between machines, so it must be safe to expose. But other
+ * software should consider it opaque, not depending on its internal structure.
+ */
+
+/**
+ * @template T
+ * @typedef {object} Subscriber
+ * @property {(publishCount?: bigint) => PublicationList<T>} subscribeAfter
+ * Internally used to get the "current" SharableSubscriptionInternals
+ * in order to make a subscription iterator that starts there.
+ * The answer is "current" in that it was accurate at some moment between
+ * when you asked and when you got the answer.
+ */
+
+/**
+ * @template T
+ * @typedef {object} Publisher
+ * A valid sequence of calls to the methods of an `IterationObserver`
+ * represents an iteration. A valid sequence consists of any number of calls
+ * to `publish` with the successive non-final values, followed by a
+ * final call to either `finish` with a successful `completion` value
+ * or `fail` with the alleged `reason` for failure. After at most one
+ * terminating calls, no further calls to these methods are valid and must be
+ * rejected.
+ * @property {(nonFinalValue: T) => void} publish
+ * @property {(completion: T) => void} finish
```

Comment:
I imagine the final and non-final values need not be the same T.

Reply:
I don't have a strong opinion on that. Shall we keep it like this for now and relax later if needed?

Reply:
No objection.

Just so you’re in the know, the drum I’ve been beating for a while now is that JS has a publisher interface we could shave and get better leverage: `<T, U>{ next(T), return(U), throw(unknown) }` (nominally `Observer`), since a utility function that can stuff iteration results into a generator can stuff them into this, and vice versa.